File: Algoritmos/01-TwoSums.cs
using System;
using System.Collections.Generic;

namespace TwoSums
{
    class Program
    {
        public static void Main (string[] args)
        {
            int[] nums = { 2, 7, 11, 15 };
            int target = 9;
            int[] result = TwoSum (nums, target);
            Console.WriteLine ("Input: nums = {0}, target = {1}", string.Join (", ", nums), target);
            Console.WriteLine ("Output: [" + result[0] + "," + result[1] + "]");

        }

        public static int[] TwoSum(int[] nums, int target)
        {
            var dic = new Dictionary<int, int>();

            for (int i = 0; i < nums.Length; i++)
            {
                if (dic.ContainsKey(nums[i]))
                    return new int[] { dic[nums[i]], i };
                else
                    dic[target - nums[i]] = i;
            }

            return new int[] { };
        }
    }
}

File: Algoritmos/01-TwoSums.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

std::vector<int> twoSum(std::vector<int>& array, int target)
{
    std::unordered_map<int, int> map;
    for (int i = 0; i < array.size(); ++i)
    {
        int t = target - array[i];
        if (map.count(t)) return { map[t], i };
        map[array[i]] = i;
    }
    return {};
}

int main()
{
    std::vector <int> array = { 2, 7, 11, 15 };
    int target = 9;

    std::vector <int> result = twoSum(array, target);

    std::cout << "Input: nums = ";

    for (int i = 0; i < array.size(); i++)
        std::cout << "[" << array[i] << "]";

    std::cout << ", target = " << target << std::endl;

    std::cout << "Output: ";

    for (int i = 0; i < result.size(); i++)
    {
        std::cout << "[" << result[i] << "]";
    }

    std::cin.get();
}
